elixir/apps/domain/lib/domain/auth/adapters/microsoft_entra/jobs.ex

```diff
@@ -0,0 +1,210 @@
+          log_effects(provider, effects)
+
+        {:error, reason} ->
+          Logger.error("Failed to sync provider",
+            provider_id: provider.id,
+            account_id: provider.account_id,
+            reason: inspect(reason)
+          )
+
+        {:error, op, value, changes_so_far} ->
+          Logger.error("Failed to sync provider",
+            provider_id: provider.id,
+            account_id: provider.account_id,
+            op: op,
+            value: inspect(value),
+            changes_so_far: inspect(changes_so_far)
+          )
+      end
+    else
+      {:error, {status, %{"error" => %{"message" => message}}}} ->
+        provider =
+          Auth.Provider.Changeset.sync_failed(provider, message)
+          |> Domain.Repo.update!()
+
+        log_sync_error(provider, "Microsoft Graph API returned #{status}: #{message}")
+
+      {:error, :retry_later} ->
+        message = "Microsoft Graph API is temporarily unavailable"
+
+        provider =
+          Auth.Provider.Changeset.sync_failed(provider, message)
+          |> Domain.Repo.update!()
+
+        log_sync_error(provider, message)
+
+      {:error, reason} ->
+        Logger.error("Failed syncing provider",
+          account_id: provider.account_id,
+          provider_id: provider.id,
+          reason: inspect(reason)
+        )
+    end
+  end
+
+  defp log_sync_error(provider, message) do
+    metadata = [
+      account_id: provider.account_id,
+      provider_id: provider.id,
+      reason: message
+    ]
+
+    if provider.last_syncs_failed >= 3 do
+      Logger.warning("Failed syncing provider", metadata)
+    else
+      Logger.info("Failed syncing provider", metadata)
+    end
+  end
+
+  defp list_membership_tuples(access_token, groups) do
+    Enum.reduce_while(groups, {:ok, []}, fn group, {:ok, tuples} ->
+      case MicrosoftEntra.APIClient.list_group_members(access_token, group["id"]) do
+        {:ok, members} ->
+          tuples = Enum.map(members, &{"G:" <> group["id"], &1["id"]}) ++ tuples
+          {:cont, {:ok, tuples}}
+
+        {:error, reason} ->
+          {:halt, {:error, reason}}
+      end
+    end)
+  end
+
+  # Map identity attributes from Microsoft Entra to Domain
+  defp map_identity_attrs(users) do
+    Enum.map(users, fn user ->
+      %{
+        "provider_identifier" => user["id"],
+        "provider_state" => %{
+          "userinfo" => %{
+            "email" => user["userPrincipalName"]
+          }
+        },
+        "actor" => %{
+          "type" => :account_user,
+          "name" => user["displayName"]
+        }
+      }
+    end)
+  end
+
+  # Map group attributes from Microsoft Entra to Domain
+  defp map_group_attrs(groups) do
+    Enum.map(groups, fn group ->
+      %{
+        "name" => "Group:" <> group["displayName"],
+        "provider_identifier" => "G:" <> group["id"]
+      }
+    end)
+  end
+
+  # Log effects of the multi transaction
+  defp log_effects(provider, effects) do
```

Comment:
Should we move it to some shared place?

Reply:
Yeah, I thought about that just today, but then forgot during some of my test debugging.  I'll try to get this DRY'd up.

Reply:
Wasn't exactly sure where to put this code, so just took a stab at it.  Let me know if this isn't a good convention or just doesn't fit in some way.